content/post/scaling-workloads/index.md

```diff
@@ -0,0 +1,116 @@
+a memory address then this will result in a higher execution time. Memory is always going to be slower
+than accessing a register. This is due to the registers being gates inside the silicon and therefore
+the lines from the 8-bit:8-bit ALU (the ALU optimized for performing an operation on 2 8-bit numbers)
+to the CPU being significantly shorter. Just from the looks of it, this program may not be that complex
+like a compiler but there is always a chance this program will not run optimally. In fact, this program
+isn't as optimal as it can be for an x86 program. The most optimal is simply this.
+
+```asm
+mov ax, 1
+add ax, 1
+```
+
+We can further optimize this if we always assume `ax` will be `1` before we run the `add` instruction.
+This results in a program which looks like this.
+
+```asm
+add ax, 1
+```
+
+This program results in a single clock cycle and takes the least amount of execution time compared
+to the other variants of this program. However, the optimization of removing the `mov` to `ax` can
+be dangerous. If `ax` is not `1` then it would result in a different result being stored into `ax`.
+This is an example of an optimization which can be dangerous. It is considered dangerous as it
+is assuming the CPU state being a certain way with no checks. Adding checks can result in degrated
+performance. This is why many program which deals with high performance may use hand written assembly.
+The compiler may introduce certain optimizations such as the one in this example that either produces
+undefined behavior or is not optimal enough. Luckily, most people do not have to deal with having to
+understand what is going on. Modern compilers like GCC and LLVM's clang are optimal enough to prevent
+most issues. However, the thing which is important is the size of the workloads, having a good CPU for
+the workload, and enough RAM for the workload. Having the right CPU also matters as well, it is better
+to have a CPU for a specific architecture instead of cross compiling. This is because you have less
+to bootstrap for a specific CPU architecture. You typically already have what you need to compile for
+your native architecture.
+
+Having enough RAM depends on what the system typically does and how much you can parallelize really are
+the most important things here. This is going to become more apparant with the world moving away from x86
+and to ARM. ARM is quite different, it is known as a RISC (Reduced Instruction Set Computing) architecture.
+x86 is considered a CISC (Complex Instruction Set Computing) architecture. x86 has many more instructions
+compared to ARM. However, many instructions in a CISC architecture are similar to each other. This means
+you may have 30 different move instructions which each are optimized for a different data type. While on ARM,
+you have just what is needed. There can be just 5 move instructions on a RISC compared to the 30 in a CISC.
```

Comment:
The RISC/CISC explanation is good but not thoroughly developed. This section and the one immediately prior are also questionably relevant to the topic of the post being process parallelization.